### ISSUE TITLE
Recurring subscription payments

### ISSUE THREAD

**Issue by 9mido** [bitcart/bitcart#174]
It would be nice to be able to allow recurring cryptocurrency payments. Also allow an active recurring paying bitcart customer to cancel their recurring subscription payments on their own to stop their recurring payments without contacting the merchant company to manually cancel the payment for them.

Also having webhooks for being able to do things like change the database your website uses like changing the membership of a customer account from paid to free when a customer cancels their recurring bitcart payment. This way, a customer can no longer access paid resources since they canceled and are now free in the database.

Same thing for a customer starting recurring bitcart subscription payments. There needs to be a way to change the database of your website so that when a customer buys a subscription, the database changes from free to paid and the customer could access paid services with their account.

Basically a complete cryptocurrency solution to what the Stripe API offers for their recurring subscription payments is desirable.

**Comment by MrNaif2018** [comment]
Hello! Thank you very much for your suggestion! We will definitely add it in next few releases! We currently have a roadmap with a rough plan of what to add in which quarter of the year, and I would move it to Invoicing improvements part. Maybe it could be implemented earlier.
It will probably be implemented store side, like, creating recurring products.

Or should it be supported at invoice level? (basically: when checking out at store POS, invoice with selected products connected is created).
Something like, instead of price it could show advanced price settings?

**Comment by 9mido** [comment]
You're welcome. Really looking forward to this as it would be a great replacement to Stripe and it would use entirely cryptocurrency that you would control yourself. 

Btcpayserver currently has https://docs.btcpayserver.org/PullPayments/ which acts like the recurring payments. They also have refunds https://docs.btcpayserver.org/Refund/. But I do not think they have anything that cancels out the recurring payments where the customer could do it themselves. 

Bitpay has subscriptions in their API and they allow you to update the status to canceled. https://bitpay.com/api/#rest-api-resources-subscriptions-resource

Stripe does it where you have a product ID like which is basically the name of your product. Each product can have multiple different subscription IDs (monthly, yearly, etc) attached to the product ID. Then they have subscription IDs where each customer gets a unique subscription ID but the subscription interval can be recurring payments that are monthly, yearly, etc. A customer could also have multiple subscription IDs (maybe 2 different product IDs with one unique subscription ID for each or 2 different subscription IDs for the same product ID). 

In Stripe, invoices have some unique statuses. https://stripe.com/docs/billing/invoices/overview. Invoices are basically bills and they usually reference a subscription ID associated with that invoice. 

When somebody cancels their active paying subscription ID payment, usually the subscription goes from active status to canceled status. When someone pays an invoice, it will be marked as paid. When the subscription ID becomes canceled, a new invoice for the next recurring month/year will not be created for the customer to pay. You can create an open invoice and you could send it to a customer to pay manually and it will remain open until they pay. On the day the next recurring subscription payment is due, the invoice will become paid if the customer has enough funds. Otherwise the system will retry paying each day until the customer pays. If the customer does not pay within a certain timeframe, the subscription can be configured to be automatically cancelled and the invoice will become unpaid / uncollectible.

A voided invoice is like voiding a written paper check but we are doing it for the bill. Uncollectible invoice means the customer may not have the funds to pay and it is impossible to contact them to have them pay it. 

Then you have things like free trials where a customer has to provide a payment method up front but will not be charged until the free trial is over. Payments could also be paused also until a different date in the future. 

Hopefully this gives you enough info on how to approach this feature.

**Comment by MrNaif2018** [comment]
Thank you for your very detailed description! I will integrate it right after finalizing BitcartCC  configurator (UI to select settings and generate script to install BitcartCC).
By the way, if you wish you can join our telegram where we discuss the new features and some important decisions. Or I can reply in this issue later on.

**Comment by 9mido** [comment]
You're welcome. And thank you for the great job. I have used stripe for quite some time now. You could always ask on IRC #stripe channel how they do things with credit cards. But developers are only there on weekdays.

**Comment by 9mido** [comment]
Here is a video about cryptocurrency subscription payments:

https://www.youtube.com/watch?v=-KLa_lDkji0

**Comment by MrNaif2018** [comment]
Thanks for this. Our roadmap is moving to invoicing improvements and other essential features. With hacktoberfest coming maybe this could be implemented soon. If someone wants to take it up, you can clarify any questions here in the issue or in our community chats.
